common/quotas/global/algorithm/requestweighted.go

```diff
@@ -0,0 +1,351 @@
+		// many times more limit keys than limiting hosts, and this will reduce cardinality more rapidly.
+		usage map[Limit]map[Identity]history
+
+		clock clock.TimeSource
+	}
+
+	// Metrics reports overall counts discovered as part of garbage collection.
+	//
+	// This is not necessarily worth maintaining verbatim, but while it's easy to
+	// collect it might give us some insight into overall behavior.
+	Metrics struct {
+		// HostLimits is the number of per-host limits that remain after cleanup
+		HostLimits int
+		// Limits is the number of cross-host limits that remain after cleanup
+		Limits int
+
+		// RemovedHostLimits is the number of per-host limits that were removed as part of this GC pass
+		RemovedHostLimits int
+		// RemovedLimits is the number of cross-host limits that were removed as part of this GC pass
+		RemovedLimits int
+	}
+
+	// Requests contains accepted/rejected request counts for a ratelimit, as part of an update operation.
+	// Only intended to be used in [RequestWeighted.Update].
+	Requests struct {
+		Accepted, Rejected int
+	}
+
+	// RequestWeighted returns aggregated RPS and distribution data for its Limit keys,
+	// based on how many requests each Identity has received.
+	RequestWeighted interface {
+		// Update load-data for this host's requests, given a known elapsed time spent accumulating this load info.
+		//
+		// Elapsed time must be non-zero, but is not otherwise constrained.
+		Update(host Identity, load map[Limit]Requests, elapsed time.Duration)
+
+		// HostWeights returns the per-[Limit] weights for all requested + known keys for this Identity,
+		// as well as the Limit's overall used RPS (to decide RPS to allow for new hosts).
+		HostWeights(host Identity, limits []Limit) (weights map[Limit]HostWeight, usedRPS map[Limit]PerSecond)
+
+		// GC can be called periodically to pre-emptively prune old ratelimits.
+		//
+		// Limit keys that are accessed normally will automatically garbage-collect themselves and old host data,
+		// and an unused limit only costs memory, so this may prove to be unnecessary.
+		GC() Metrics
+	}
+
+	Config struct {
+		// How much each update should be weighted vs prior data.
+		// Must be between 0 and 1, recommend starting with 0.5 (4 updates until data has <10% influence)
+		NewDataWeight dynamicconfig.FloatPropertyFn
+
+		// Expected rate of updates.  Should match the cluster's config for how often limiters check in.
+		UpdateRate dynamicconfig.DurationPropertyFn
+
+		// How many UpdateRate periods can pass without new information before considering a host-limit's
+		// RPS "probably inactive", rather than simply delayed.
+		//
+		// Should always be 1 or larger.  Unsure about a good default (try 2+?), but larger numbers mean smoother
+		// behavior but longer delays on adjusting to hosts that have disappeared or stopped receiving requests.
+		MaxMissedUpdates dynamicconfig.FloatPropertyFn
```

Comment:
Yea, I was considering this too.  I've gotta re-think it to make sure it handles everything I need, but it might / at worst I could probably just divide this by `UpdateRate` and infer the same thing (unsure if that'll be clearer or worse tho)

Reply:
yep, durations everywhere feel pretty good.  gonna stick with that.